Automatically extract page contributor information during web build

VojtechStep opened the pull request:
The contributor list is sorted by number of commits. Any suggestions for wording and visual style are appreciated. There is a new system for attributing work to contributors, which is now opt-in; add yourself to `scripts/contributors_data.toml` if you want to be mentioned on the website, or have someone do it for you.

See the latest version here: https://unimath.vojtechstep.eu

## Homepage
The logo now properly changes color on dark themes. The font size is bigger, and some of the window size calculations were changed so that the content doesn't overlap the right sidebar, and that sidebar is hidden on smaller screens where it's not useful. It's also hidden when printing a page with `Ctrl+P`, which should make it more usable.

![image](https://github.com/UniMath/agda-unimath/assets/15523887/a3a14618-42c3-49a7-affb-17cb418f3e05)

## Page attribution

The "Content created by:" line is only included in Agda source files, and the Created/Modified and Recent changes sections are only included in Agda source files and guides in the Overview section.

![image](https://github.com/UniMath/agda-unimath/assets/15523887/9f807028-98a8-4f74-9def-87b0b134a24f)

## Recent changes
![image](https://github.com/UniMath/agda-unimath/assets/15523887/8d7efa6b-ee0c-4f40-8db6-04052736bad7)

## Updated list of contributors

Now including co-authors; visible names, links to github and association between visible names and git usernames can be customized via the `scripts/contributors_data.toml` file

![image](https://github.com/UniMath/agda-unimath/assets/15523887/db6616bd-2601-4a82-a4a5-24d9c2b40b99)

## Proper syntax highlighting with dark theme support

![image](https://github.com/UniMath/agda-unimath/assets/15523887/f08fd5a7-2a78-4950-ad9a-63ecb98b57bc)

![image](https://github.com/UniMath/agda-unimath/assets/15523887/e5886705-f46a-42dd-b8e4-5dc0e7d7e8d3)

## Next steps

#777 and #459 are still open; I'm still not convinced about Egbert's idea of listing changed files for every contributor.

Fixes #439
Fixes #507

EgbertRijke commented:
Would it be possible/desirable to display this information immediately below the title of the page?

Don't we want to sort contributors by last name, if a first and last name are available?

Would it be possible to also display the date that the file was created, and perhaps the last date that the file was edited?

EgbertRijke commented:
Also: Awsome!!

VojtechStep commented:
I moved the contributors to the top.

VojtechStep commented:
I added the "Created" and "Modified" fields, and now's the time when I'll need anyone's aesthetic help to make it look reasonable 😅.
I can also offer the following alternative date formats:
![image](https://github.com/UniMath/agda-unimath/assets/15523887/4a0d20bd-aad1-44e9-8ca5-8c4e0e209524)
Personally I prefer the current one, `%as`, i.e. `YYYY-MM-DD`.

VojtechStep commented:
I added the "Recent changes" section to the bottom, I'll need ideas to style that as well. I also don't think it should be a problem to add links to the corresponding commits in the GitHub repo.

That's everything from me for today.

VojtechStep commented:
I also didn't address the contributor sorting; TBH I like it sorted by number of commits (if only for the fact that git does that automatically for me and I don't have to do any further processing 😅).

EgbertRijke commented:
> I added the "Created" and "Modified" fields, and now's the time when I'll need anyone's aesthetic help to make it look reasonable 😅.
> I can also offer the following alternative date formats:
> ![image](https://user-images.githubusercontent.com/15523887/268125045-4a0d20bd-aad1-44e9-8ca5-8c4e0e209524.png)
> Personally I prefer the current one, `%as`, i.e. `YYYY-MM-DD`.

Yes, just the date is fine

EgbertRijke commented:
> I also didn't address the contributor sorting; TBH I like it sorted by number of commits (if only for the fact that git does that automatically for me and I don't have to do any further processing 😅).

Sure, that works for now. But that also sounds like we can't manually add names because this script will always rewrite the contributor attribution?

EgbertRijke commented:
Instead of just "Modified" I would say "Last modified"

VojtechStep commented:
> > I also didn't address the contributor sorting; TBH I like it sorted by number of commits (if only for the fact that git does that automatically for me and I don't have to do any further processing 😅).
> 
> Sure, that works for now. But that also sounds like we can't manually add names because this script will always rewrite the contributor attribution?

That is correct, all of this information is automatically extracted from git metadata with no option to tweak it from within the file whatsoever.

EgbertRijke commented:
BTW it doesn't look like its sorted by number of commits, but it's sorted by first name. We had this problem before: with this sorting it displays capitalized names before uncapitalized names. It looks wrong to me.

VojtechStep commented:
> BTW it doesn't look like its sorted by number of commits, but it's sorted by first name. We had this problem before: with this sorting it displays capitalized names before uncapitalized names. It looks wrong to me.

Nope, it's certainly sorted by number of commits - the names are explicitly generated with the `-n` flag to `git shortlog` for it to be so, and some file have "Egbert Rijke and Fredrik Bakke", while others have "Fredrik Bakke and Egbert Rijke".

VojtechStep commented:
It is possible that multiple users have the same number of commits (as is the case in the Order theory file, where Jonathan, Victor and Louis all have 1 commit); in which case I assume it sorts alphabetically

EgbertRijke commented:
> > BTW it doesn't look like its sorted by number of commits, but it's sorted by first name. We had this problem before: with this sorting it displays capitalized names before uncapitalized names. It looks wrong to me.
> 
> Nope, it's certainly sorted by number of commits - the names are explicitly generated with the `-n` flag to `git shortlog` for it to be so, and some file have "Egbert Rijke and Fredrik Bakke", while others have "Fredrik Bakke and Egbert Rijke".

Darn! I have to find those files and work on them harder!

EgbertRijke commented:
There is one more issue I would like to know. Sometimes people change their names. When they do, and update their info on github, do the updated names show up automatically on our website? Or does it show their names at the time they committed the work? I would not like it if our website accidentally outs those people.

VojtechStep commented:
All this information is from git, not GitHub; git has no concept of identity beyond username and email, which are associated to every commit at the time of authoring, and they are part of the commit data, so changing it would require rewriting history. I would like to show GitHub usernames instead, but we'd need some sort of mapping from git usernames to GitHub usernames. I haven't thought this through, I'll take a look at it either tomorrow or on Sunday.

EgbertRijke commented:
Ok, let's think this through carefully.

VojtechStep commented:
I summarized the result of our discussion with @EgbertRijke in the description of the last commit, let me paste it here as well:


After some discussion with Egbert, we came to the conclusion that we
should make it possible for contributors to the library to choose how
they are presented on the website, without that presentation being
immediately linked to their GitHub account(s), and without needing to
rewrite the commit history in situations where the identity associated
with past commits is no longer the one a contributor wants to be
identified with.

The solution we came up with is to have a collection of contributor
identities, not a priori linked to any other service, where
individuals can set their preferred display name and "claim" git
usernames. When information obtained from the git log is
presented (currently the "Created by" page header and "Recent changes"
footer of pages), this table is consulted at website build-time to
display the preferred name.

If a contributor chooses not to be presented on the website, they can
be left out of the table.

The next step is to use this table to generate the CONTRIBUTORS.md
file from git log as well, instead of relying on GitHub APIs.

EgbertRijke commented:
For recent changes, perhaps the following format would be cleaner:

`Date. Author (and coathors?). _Short commit title_ (#PR). Short commit hash.`

VojtechStep commented:
> For recent changes, perhaps the following format would be cleaner:
> 
> `Date. Author (and coathors?). _Short commit title (#PR)_. Short commit hash.`

Thanks for pointing out the missing coauthors, they should be properly included in the Recent changes section now. I tried you suggested formatting; I think it's a step in the right direction, although I'm not a fan of the hash at the end. I'll try and play with it a bit more on Sunday.
![image](https://github.com/UniMath/agda-unimath/assets/15523887/b013d373-43d9-4eec-abd6-f23e3ce9922e)

EgbertRijke commented:
Well, since these titles are mostly PR titles (unless the commits are prehistoric) I think it makes sense to link to the specific PR. It is an easy way to get more information about the specific change that is listed here under recent changes.

EgbertRijke commented:
Perhaps if we don't display the hash, the formatting could be cleaned up as follows:

`Date. Authors and coauthors. _Title_. [#PR](link-to-pr)`

fredrik-bakke commented:
This makes it very similar to how we cite references currently, which is a nice bonus to consistency :)

VojtechStep commented:
I'd rather not try to be smart and parse the PR from the commit message, but instead make a link to the specific commit (which we can do, since that's derived from the hash), and there GitHub adds the link for us - for example this a link to the last commit above: https://github.com/UniMath/agda-unimath/commit/bfab47422b6f0110462fba1081ebd786257f38fc

EgbertRijke commented:
> I'd rather not try to be smart and parse the PR from the commit message, but instead make a link to the specific commit (which we can do, since that's derived from the hash), and there GitHub adds the link for us - for example this a link to the last commit above: [bfab474](https://github.com/UniMath/agda-unimath/commit/bfab47422b6f0110462fba1081ebd786257f38fc)
> 
> edit: Github shortens the link, it looks like this `https://github.com/UniMath/agda-unimath/commit/bfab47422b6f0110462fba1081ebd786257f38fc`

Sounds good!

EgbertRijke commented:
On the other hand, the link to pull request #XYZ seems to be just

https://github.com/UniMath/agda-unimath/pull/XYZ

So it looks like "trying not to be smart" might actually require some effort in this case

VojtechStep commented:
> On the other hand, the link to pull request #XYZ seems to be just
> 
> `https://github.com/UniMath/agda-unimath/pull/XYZ`
> 
> So it looks like "trying not to be smart" might actually require some effort in this case

"Trying not to be smart" in this case means not parsing the commit message at all, and instead use the commit hash we get from git 🤷

EgbertRijke commented:
> > On the other hand, the link to pull request #XYZ seems to be just
> > `https://github.com/UniMath/agda-unimath/pull/XYZ`
> > So it looks like "trying not to be smart" might actually require some effort in this case
> 
> "Trying not to be smart" in this case means not parsing the commit message at all, and instead use the commit hash we get from git 🤷

Lol, ok, I shouldn't try to be smart because that's not working out so well.

EgbertRijke commented:
Btw what stops us from displaying the full history of a file somewhere?

EgbertRijke commented:
...or what stops us from turning it into a full wiki with a version of Agda that works in the browser? 🥹

VojtechStep commented:
The version with links, I like it:
![image](https://github.com/UniMath/agda-unimath/assets/15523887/645c1299-e92d-4d8e-ae2b-a25e14e390a8)

VojtechStep commented:
> Btw what stops us from displaying the full history of a file somewhere?

Well, you can already see a file's history on GitHub, why put it on our website too?

EgbertRijke commented:
I love it! Would it look even better if only the PR number would be a link?

EgbertRijke commented:
If it has three or more coathors, is the list comma-separated or does it have "and" between all the coathors?

VojtechStep commented:
> I love it! Would it look even better if only the PR number would be a link?

Well, at that point I might as well just parse the number 😅. I'll see what I can do in the bus, but for now I'm signing out ✌️

EgbertRijke commented:
In that case leave it the way it is. It looks very good!

VojtechStep commented:
> If it has three or more coathors, is the list comma-separated or does it have "and" between all the coathors?

It uses the same logic as the top-of-file attribution, so see the screenshot in the PR description

fredrik-bakke commented:
For what it's worth, I kind of like the way this format downplays the title of the PRs 😶

VojtechStep commented:
I added a link to a personal deployment of the unimath website, so that you can browse it in full in your spare time.

VojtechStep commented:
I added a little roadmap; the current plan is to look over the website issues, see what I can do during the 10 hours I have on the bus, and then open the PR with whatever I managed.

EgbertRijke commented:
I just had a look at the website you're making. One thing I noticed is that to me it seems less appropriate to have the content attributions and dates on the home page and any nonformalization pages (community/guides/art). What do you think about it? Would it cause you a headache to remove them?

VojtechStep commented:
I can remove them pretty easily, I agree that the attributions probably don't make much sense there. I would keep the dates and recent changes though, since the files are changed from time to time; what do you think?

EgbertRijke commented:
On the homepage I think the dates and recent changes are not appropriate either. That's where people enter the page, and it should look clean and inviting. Actually I already don't like it that it has this strange long list of all files, but apparently that is an mdbook artefact that we must have (or is it?).

On any of the other pages, I'm fine with having the dates and recent changes.

EgbertRijke commented:
I also don't quite know why the list of contributors is listed under the long list of all files, where no one will find it. To me it all looks very clumsy:)

VojtechStep commented:
You're correct that the structure of the homepage w.r.t all the links is required by mdbook - a page appears on the webpage iff it is listed in this summary. If you want I can try to make that part collapsible or something, but the mdbook docs explicitly say "its formatting is very strict and must follow the structure outlined below. Any element not specified below is likely to be ignored at best, or may cause an error", so I'm not making any promises.

I'll remove the git decorations from it.

EgbertRijke commented:
We could also get rid of the logo. Or what about a cool logo like this:

![agda-unimath-star-trek-extra-bold-bt](https://github.com/UniMath/agda-unimath/assets/1252282/797bf1d7-2a21-426c-b037-8029d3bec321)

VojtechStep commented:
Oh wait no, actually home and summary are different files, and HOME just includes SUMMARY; I can try removing that tomorrow and see what it does.

VojtechStep commented:
> We could also get rid of the logo

And you say that now, after I made it play nice with the dark themes 😭

EgbertRijke commented:
Sorry!

EgbertRijke commented:
> Oh wait no, actually home and summary are different files, and HOME just includes SUMMARY; I can try removing that tomorrow and see what it does.

It also includes contributors for no reason

EgbertRijke commented:
Perhaps that summary file should be accessible through the navigation menu on the left, but it should not be included in the home page.

EgbertRijke commented:
> > We could also get rid of the logo
> 
> And you say that now, after I made it play nice with the dark themes 😭

It does look a little bit better on the new website ❤️

EgbertRijke commented:
On Tuesday next week (9 days from now) I'm demonstrating the library at a summer school. Would it be possible to merge this PR by then?

VojtechStep commented:
I'd hope to have it done by tonight or tomorrow

VojtechStep commented:
> It also includes contributors for no reason

@EgbertRijke Just to make sure, does this mean you want me to remove the list of contributors from the homepage?

EgbertRijke commented:
Not from the homepage altogether. The list of contributors should still be found under "Community".

There is just no reason to have it on HOME.md after the inclusion of SUMMARY.md where noone will find it.

VojtechStep commented:
I meant specifically on the "HOME" page; I agree that it doesn't make sense to have them listed under the summary, but now I'm removing the summary from the homepage, and I wasn't sure if I should keep the contributors there or not

fredrik-bakke commented:
> On Tuesday next week (9 days from now) I'm demonstrating the library at a summer school. Would it be possible to merge this PR by then?

Are you presenting at [Interactions of Proof Assistants and Mathematics](https://itp-school-2023.github.io/) by chance? (I won't be there myself, but I can only regret that.)

EgbertRijke commented:
Yup

EgbertRijke commented:
> I meant specifically on the "HOME" page; I agree that it doesn't make sense to have them listed under the summary, but now I'm removing the summary from the homepage, and I wasn't sure if I should keep the contributors there or not

I think they can be removed. It is clear enough if the contributors page is under Community, and the home page should just be clean and welcoming.

VojtechStep commented:
The PR is now ready for review. I added a summary of the changes and appropriate screenshots to the description. Please take a look and tell me if you see anything wrong with the updated website. I rebased on master, now I'm building the website locally, and the latest version should appear under the link in the description shortly (currently it's the latest version before the rebase).

EgbertRijke commented:
Instead of the terse "Created: 2022-03-11; Last modified: 2023-09-13" could we use the format

Created on 2022-03-11.
Last modified on 2023-09-13.

VojtechStep commented:
Sure; including the line break?

EgbertRijke commented:
Yes, I think it will look better with it.

EgbertRijke commented:
The following names need corrections:

[Elif Uskuplu](https://github.com/ElifUskuplu)
[Victor Blanchi](https://github.com/VictorBlanchi)
[Ian Ray](https://github.com/IanRay11)
[Ivan Kobe](https://github.com/ivankobe)
[Daniel Gratzer](https://github.com/jozefg)
[Maša Žaucer](https://github.com/masazaucer)

VojtechStep commented:
Thank you; I chose the display names that appeared on the existing contributors page, if these are more appropriate I will change them.

EgbertRijke commented:
Can I launch it?:)

VojtechStep commented:
If you feel like it 👀

EgbertRijke commented:
Let's gooooooo 🚀

EgbertRijke commented:
Thank you for your amazing work!

VojtechStep commented:
Now let's cross our fingers that the website build runs correctly in the CI 😅

EgbertRijke commented:
It would be so funny if you made the entire website redirect [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

EgbertRijke commented:
> Now let's cross our fingers that the website build runs correctly in the CI 😅

You shouldn't have said that:)

VojtechStep commented:
macOS doesn't have `cp -T`? How do you guys live?

EgbertRijke commented:
Who said we live?

VojtechStep commented:
I think I have to make a new PR for the fix

EgbertRijke commented:
Lol, get some sleep

VojtechStep commented:
If that next PR doesn't fix it then I'm leaving it for tomorrow

VojtechStep commented:
And I just saw that the checkout action does a `--depth=1`, so I'll also need to fix that, but that'll have to wait until tomorrow

EgbertRijke commented:
Actually your fix doesn't show the names and the recent changes anymore

VojtechStep commented:
Hm, you're right, it looks like the git metadata option didn't get enabled

VojtechStep commented:
That's so weird, the CI straight up doesn't run the script. At least the contributor list is now fixed. I'll investigate more today.

EgbertRijke commented:
When I try to build the website locally I get this (if it is any help to you):

```
Traceback (most recent call last):
  File "/Users/egbertrijke/Repositories/agda-unimath/./scripts/generate_contributors.py", line 7, in <module>
    from utils.contributors import CONTRIBUTORS_FILE, parse_contributors_file, sorted_authors_from_raw_shortlog_lines
  File "/Users/egbertrijke/Repositories/agda-unimath/scripts/utils/contributors.py", line 3, in <module>
    import tomli
ModuleNotFoundError: No module named 'tomli'
make: *** [CONTRIBUTORS.md] Error 1
```

VojtechStep commented:
I added a Python dependency to requirements.txt, I should have mentioned that, please run `python3 -m pip install -r scripts/requirements.txt` again. The thing is that if the script failed, we'd get an error log from mdbook and the action would fail. As it stands it looks like mdbook never runs the preprocessor at all. I'm experimenting with the action in my repo so that we don't have to speculatively merge PRs

EgbertRijke commented:
It seems that one new module called `tomli` is required. We should make a habit of updating the [install guide](https://unimath.github.io/agda-unimath/HOWTO-INSTALL.html#setup-for-contributors) hen we add new dependencies to the repository. Is `tomli` the only new requirement?

VojtechStep commented:
What would the update say? We already say that the developer needs to run the above command, you just need to rerun it when the list of dependencies changes.

EgbertRijke commented:
It should say what the script does, i.e., what it installs when you run it. We should not ask people to trust us with our script if we don't tell them what it does:)

VojtechStep commented:
#786 should both fix the CI and document the new dependency
